### PR TITLE
Rocksdb Block Storage

### PR DESCRIPTION
**Pull request by ArushiSinghal**
### Description of the Change
Implement RocksDB BlockStorage class

### CODE REVIEW

**Comment by ArushiSinghal** [comment]
@MBoldyrev @lebdron This is a draft of the RocksDB block storage implementation. Adding tests for the class now.

**Comment by ArushiSinghal** [comment]
@MBoldyrev @lebdron Please look and verify the "fetch" function.

File: vendor/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+add_library(rocksdb_block_storage
```
Comment (lebdron) [inline vendor/CMakeLists.txt:1]: This is better placed in irohad/ametsuchi/impl directory, together with other implementations of block storage.

File: vendor/impl/rocksdb_block_storage.hpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		void forEach(FunctionType function) const override;
+
+	private:
+		std::string const name = (boost::filesystem::temp_directory_path()
+				/ boost::filesystem::unique_path())
+			.string();
+
+		rocksdb::DB *db;
+		rocksdb::Options options;
+		rocksdb::Status status;
```
Comment (lebdron) [inline vendor/impl/rocksdb_block_storage.hpp:39]: No need to make options and status and class fields. Use variables in methods directly.

File: vendor/impl/rocksdb_block_storage.hpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	// RocksDB Block Constructor constructor
+	public:
+		RocksdbBlockStorage() {
```
Comment (lebdron) [inline vendor/impl/rocksdb_block_storage.hpp:43]: Instead, pass a `rocksdb::DB *` to constructor if it was opened successfully.

File: vendor/impl/rocksdb_block_storage_factory.cpp [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#include "vendor/impl/rocksdb_block_storage.hpp"
+
+#include "vendor/impl/rocksdb_block_storage_factory.hpp"
+
+std::unique_ptr<iroha::ametsuchi::BlockStorage> RocksdbBlockStorageFactory::create() {
```
Comment (lebdron) [inline vendor/impl/rocksdb_block_storage_factory.cpp:5]: Check the path provider approach used in `FlatFileBlockStorageFactory`.

**Review by lebdron — CHANGES_REQUESTED** [review]

File: test/module/vendor/CMakeLists.txt [hunk 5]
```diff
@@ -12,3 +12,19 @@ addtest(tbb_test tbb_test.cpp)
...
     TBB::tbb
     )
+
+addtest(rocksdb_test rocksdb_test.cpp)
+target_link_libraries(rocksdb_test
+    Boost::filesystem
+    RocksDB::rocksdb
+    )
+
+addtest(rocksdb_block_storage_test rocksdb_block_storage_test.cpp)
```
[the remark this answers is not in the document]
Comment (ArushiSinghal) [reply test/module/vendor/CMakeLists.txt:22]: Will do. Noted.

[the remark this answers is not in the document]
Comment (ArushiSinghal) [reply test/module/vendor/CMakeLists.txt:22]: ok. So I'll move both test and impl to respective ametsuchi directory.

**Comment by lebdron** [comment]
Also, check that all commits are signed off.

**Comment by ArushiSinghal** [comment]
> Also, check that all commits are signed off.

I'll check and squash the commits into one

File: vendor/impl/rocksdb_block_storage_factory.cpp (hunk 4 above)
Comment (ArushiSinghal) [reply vendor/impl/rocksdb_block_storage_factory.cpp:5]: Done.

File: vendor/impl/rocksdb_block_storage.hpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#include "common/bind.hpp"
+#include "backend/protobuf/block.hpp"
+#include "ametsuchi/block_storage_factory.hpp"
+#include "backend/protobuf/proto_block_factory.hpp"
+#include "interfaces/iroha_internal/abstract_transport_factory.hpp"
+#include "logger/logger.hpp"
+
+using iroha::operator|;
+
+#ifndef ROCKSDB_BLOCK_STORAGE_HPP
```
[the remark this answers is not in the document]
Comment (ArushiSinghal) [reply vendor/impl/rocksdb_block_storage.hpp:15]: noted

File: vendor/impl/rocksdb_block_storage.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
+#include "rocksdb_block_storage.hpp"
```
[the remark this answers is not in the document]
Comment (ArushiSinghal) [reply vendor/impl/rocksdb_block_storage.cpp:1]: Noted.

File: vendor/CMakeLists.txt (hunk 1 above)
Comment (ArushiSinghal) [reply vendor/CMakeLists.txt:1]: noted

**Comment by ArushiSinghal** [comment]
@lebdron comments were noted and pull request sent.

**Comment by kamilsa** [comment]
Similar things are implemented in #1084

**Pull request closed without merging**